Handle despawn

superdump opened the pull request:
Including a couple of fun little demos that I modified to test that it works.

superdump commented on src/physics/systems.rs:
```diff
@@ -182,3 +190,86 @@ pub fn sync_transform_system(
+    // FIXME: How should joints be removed?
+    // for entity in &joints_removed.clone() {
+    //     if let Some((joint, entity1, entity2)) = entity_maps.joints.get(entity) {
+    //         ??? joints.remove(*joint, &mut bodies);
```
I don't know what to do about joints as `JointSet` does not have a `remove()` function.

superdump replied:
Something like this could have worked but `RigidBody.joint_graph_index` and `JointSet.remove_rigid_body()` are crate-private:
```rust
    for entity in &joints_removed.clone() {
        if let Some((joint, entity1, entity2)) = entity_maps.joints.get(entity) {
            if let Some(rbh1) = entity_maps.bodies.get(entity1) {
                if let Some(rb1) = bodies.get(*rbh1) {
                    joints.remove_rigid_body(rb1.joint_graph_index, &mut bodies);
                }
            }
            if let Some(rbh2) = entity_maps.bodies.get(entity2) {
                if let Some(rb2) = bodies.get(*rbh2) {
                    joints.remove_rigid_body(rb2.joint_graph_index, &mut bodies);
                }
            }
            joints_removed.remove(entity);
            entity_maps.joints.remove(entity);
        }
    }
```

sebcrozet replied:
Joint removal will have to be implemented in Rapier first.

superdump replied:
Works now with rapier 0.3.1. :)

sebcrozet reviewed (changes requested):
Thank you for this PR!

Having to maintain an `EntityMap` resource is very unfortunate. If I understand correctly, the reason why we need this is that once the component has been removed from the entity, there is no way of recovering it from some kind of Bevy queries?

Also I feel like the despawn system could be much simpler. Is there anything the following would not cover (besides joints of course):

```rust
/// System responsible for removing joints, colliders, and bodies that have
/// been removed from the scene
pub fn destroy_body_and_collider_system(
    mut commands: Commands,
    mut bodies: ResMut<RigidBodySet>,
    mut colliders: ResMut<ColliderSet>,
    mut joints: ResMut<JointSet>,
    mut entity_maps: ResMut<EntityMaps>,
    collider_query: Query<(Entity, &ColliderHandleComponent)>,
    joint_query: Query<(Entity, &JointHandleComponent)>,
    body_query: Query<(Entity, &RigidBodyHandleComponent)>,
) {
    // Components removed before this system
    let bodies_removed = body_query
        .removed::<RigidBodyHandleComponent>()
        .into_iter();
    let colliders_removed = collider_query
        .removed::<ColliderHandleComponent>();
    let joints_removed = joint_query
        .removed::<JointHandleComponent>();

    for entity in bodies_removed {
        if let Some(body) = entity_maps.bodies.get(&entity) {
            bodies.remove(*body, &mut colliders, &mut joints);
            entity_maps.bodies.remove(&entity);
            
            // Removing a body also removes its colliders and joints. If they were
            // not also removed then we must remove them here.
            commands.remove_one::<ColliderHandleComponent>(*entity);
            commands.remove_one::<JointHandleComponent>(*entity);
        }
    }

    for entity in colliders_removed {
        if let Some(collider) = entity_maps.colliders.get(entity) {
            colliders.remove(*collider, &mut bodies);
            entity_maps.colliders.remove(&entity);
        }
    }

    // FIXME: How should joints be removed?
    // for entity in &joints_removed.clone() {
    //     if let Some((joint, entity1, entity2)) = entity_maps.joints.get(entity) {
    //         ??? joints.remove(*joint, &mut bodies);
    //         joints_removed.remove(entity);
    //         entity_maps.joints.remove(entity);
    //     }
    // }
}
```

superdump commented:
I think `commands.remove_one::<T>(entity)` will return an error result if the entity does not have such a component. We could do `let _ = commands.remove_one::<T>(entity);` to ignore the result entirely. That said, I liked having the checks and warnings printed because it lets you know when something is leaked and until we fix that, that could be useful. What do you think?

sebcrozet commented:
> I think `commands.remove_one::<T>(entity)` will return an error result if the entity does not have such a component. We could do `let _ = commands.remove_one::<T>(entity);` to ignore the result entirely.

This sounds reasonable. If the component isn't there that means it has already been removed somehow, and we should be OK with that since we would have removed it anyways.

> That said, I liked having the checks and warnings printed because it lets you know when something is leaked and until we fix that, that could be useful. What do you think?

In what case exactly could we end up with a leak?

superdump commented:
> > That said, I liked having the checks and warnings printed because it lets you know when something is leaked and until we fix that, that could be useful. What do you think?
> 
> In what case exactly could we end up with a leak?

For now for example we can’t handle removal of joints without changes to rapier. Arguably we should just fix that though. I’ll make a PR over there soon unless someone beats me to it.

sebcrozet commented:
> For now for example we can’t handle removal of joints without changes to rapier. Arguably we should just fix that though. I’ll make a PR over there soon unless someone beats me to it.

I implemented the joint removal in the version 0.3.1 of Rapier. So if this was the only case where we would end up with a leak, I'd suggest to use the simplified despawn system without the hash-maps and the `eprintln!()`.

superdump commented:
I merged the bevy 0.3 changes from #24, so merge that first. I simplified as much as I could but the `HashMap`s had to remain as the components are gone already so we can't use `query.get_component::<Component>(entity)` to look them up. And I added two joint despawn examples to test that they work too.

sebcrozet commented:
All good, thanks!
